### PR TITLE
add irep and lua-lang virtual dependency

### PR DESCRIPTION
**Pull request by RikkiButler20**
Add new package, irep.

### CODE REVIEW

File: var/spack/repos/builtin/packages/irep/package.py [hunk 1]
```diff
@@ -0,0 +1,48 @@
+# Copyright 2013-2021 Lawrence Livermore National Security, LLC and other
+# Spack Project Developers. See the top-level COPYRIGHT file for details.
+#
+# SPDX-License-Identifier: (Apache-2.0 OR MIT)
+
+# ----------------------------------------------------------------------------
```
Comment (tldahlgren) [inline var/spack/repos/builtin/packages/irep/package.py]: Please remove the boiler plate at lines 6-21

File: var/spack/repos/builtin/packages/irep/package.py [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+from spack import *
+
+
+class Irep(Package):
+    """IREP is a tool that enables mixed-language simulation codes to use a
+    common, Lua-based format for their input decks. Essentially, the input
+    format is a set of tables -- Lua's one (and only?) data structure."""
+
+    # FIXME: Add a proper url for your package's homepage here.
+    homepage = "https://irep.readthedocs.io/en/latest/"
```
Comment (tldahlgren) [inline var/spack/repos/builtin/packages/irep/package.py]: You shouldn't need to specify the language/latest path.

```suggestion
    homepage = "https://irep.readthedocs.io"
```

File: var/spack/repos/builtin/packages/irep/package.py [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+
+from spack import *
+
+
+class Irep(Package):
+    """IREP is a tool that enables mixed-language simulation codes to use a
+    common, Lua-based format for their input decks. Essentially, the input
+    format is a set of tables -- Lua's one (and only?) data structure."""
+
+    # FIXME: Add a proper url for your package's homepage here.
```
Comment (tldahlgren) [inline var/spack/repos/builtin/packages/irep/package.py]: Please remove this boiler plate

File: var/spack/repos/builtin/packages/irep/package.py [hunk 4]
```diff
@@ -0,0 +1,48 @@
...
+    homepage = "https://irep.readthedocs.io/en/latest/"
+    url      = "https://github.com/LLNL/irep"
+
+    # FIXME: Add a list of GitHub accounts to
+    # notify when the package is updated.
+    # maintainers = ['github_user1', 'github_user2']
```
Comment (tldahlgren) [inline var/spack/repos/builtin/packages/irep/package.py]: Consider adding maintainers (e.g., 'tgamblin') but check with them.  Also remove the boilerplate (i.e., lines 35- 36).

File: var/spack/repos/builtin/packages/irep/package.py [hunk 5]
```diff
@@ -0,0 +1,48 @@
...
+#
+#     spack edit irep
+#
+# See the Spack documentation for more information on packaging.
+# ----------------------------------------------------------------------------
+
+from spack import *
+
+
+class Irep(Package):
```
Comment (tldahlgren) [inline var/spack/repos/builtin/packages/irep/package.py]: Is there a reason for not inheriting from `MakefilePackage`?

(https://spack.readthedocs.io/en/latest/build_systems/makefilepackage.html)

**Comment by tldahlgren** [comment]
@RikkiButler20 Oops.  Did not see that the PR is a `WIP` so I won't add more comments until it's ready.

**Comment by RikkiButler20** [comment]
No problem. I forgot to put the WIP tag. Thanks for the comments.

File: var/spack/repos/builtin/packages/irep/package.py (hunk 4 above)
Comment (tgamblin) [reply var/spack/repos/builtin/packages/irep/package.py]: Add @tomstitt and @kennyweiss here

**Review by tgamblin — APPROVED** [review]

**Comment by tgamblin** [comment]
Looks like CI is still failing.

**Review by tgamblin — CHANGES_REQUESTED** [review]
This needs to build from scratch for both of these:

```
spack install irep ^lua
spack install irep ^lua-luajit
```

Currently, the first one works, but not the second -- because `lua-luajit` has no `lua` executable.  It actually works on LC machines because `lua` is in the default environment there, but if you try it on a plain old `spack/centos7` image that does *not* have `lua` installed, you get:

```
==> Error: ProcessError: Command exited with status 1:
    'cmake' '-G' 'Unix Makefiles' '-DCMAKE_INSTALL_PREFIX:STRING=/root/spack/opt/spack/linux-centos7-haswell/gcc-4.8.5/irep-1.0.0-j75bevag4sdypalugcywsyz5ijja6mly' '-DCMAKE_BUILD_TYPE:STRING=RelWithDebInfo' '-DCMAKE_INTERPROCEDURAL_OPTIMIZATION:BOOL=OFF' '-DCMAKE_VERBOSE_MAKEFILE:BOOL=ON' '-DCMAKE_INSTALL_RPATH_USE_LINK_PATH:BOOL=OFF' '-DCMAKE_INSTALL_RPATH:STRING=/root/spack/opt/spack/linux-centos7-haswell/gcc-4.8.5/irep-1.0.0-j75bevag4sdypalugcywsyz5ijja6mly/lib;/root/spack/opt/spack/linux-centos7-haswell/gcc-4.8.5/irep-1.0.0-j75bevag4sdypalugcywsyz5ijja6mly/lib64;/root/spack/opt/spack/linux-centos7-haswell/gcc-4.8.5/lua-luajit-2.0.5-hr2tpznfx45ryrw2dssblxgscrqi3r7k/lib' '-DCMAKE_PREFIX_PATH:STRING=/root/spack/opt/spack/linux-centos7-haswell/gcc-4.8.5/lua-luajit-2.0.5-hr2tpznfx45ryrw2dssblxgscrqi3r7k;/root/spack/opt/spack/linux-centos7-haswell/gcc-4.8.5/cmake-3.19.7-wrvfl7qvojl6pquvuabpcnq2wh7oykah' '/tmp/root/spack-stage/spack-stage-irep-1.0.0-j75bevag4sdypalugcywsyz5ijja6mly/spack-src'

1 error found in build log:
     9     -- Detecting C compile features - done
     10    -- Detecting Fortran compiler ABI info
     11    -- Detecting Fortran compiler ABI info - done
     12    -- Check for working Fortran compiler: /root/spack/lib/spack/env/gcc
           /gfortran - skipped
     13    -- Checking whether /root/spack/lib/spack/env/gcc/gfortran supports 
           Fortran 90
     14    -- Checking whether /root/spack/lib/spack/env/gcc/gfortran supports 
           Fortran 90 - yes
  >> 15    CMake Error at /root/spack/opt/spack/linux-centos7-haswell/gcc-4.8.5
           /cmake-3.19.7-wrvfl7qvojl6pquvuabpcnq2wh7oykah/share/cmake-3.19/Modu
           les/FindPackageHandleStandardArgs.cmake:218 (message):
     16      Could NOT find Lua (missing: LUA_LIBRARIES LUA_INCLUDE_DIR)
     17    Call Stack (most recent call first):
     18      /root/spack/opt/spack/linux-centos7-haswell/gcc-4.8.5/cmake-3.19.7
           -wrvfl7qvojl6pquvuabpcnq2wh7oykah/share/cmake-3.19/Modules/FindPacka
           geHandleStandardArgs.cmake:582 (_FPHSA_FAILURE_MESSAGE)
     19      /root/spack/opt/spack/linux-centos7-haswell/gcc-4.8.5/cmake-3.19.7
           -wrvfl7qvojl6pquvuabpcnq2wh7oykah/share/cmake-3.19/Modules/FindLua.c
           make:232 (FIND_PACKAGE_HANDLE_STANDARD_ARGS)
     20      CMakeLists.txt:13 (find_package)
     21    

See build log for details:
  /tmp/root/spack-stage/spack-stage-irep-1.0.0-j75bevag4sdypalugcywsyz5ijja6mly/spack-build-out.txt
```

The key part is this:

```
     16      Could NOT find Lua (missing: LUA_LIBRARIES LUA_INCLUDE_DIR)
```

I was able to get this working by adding some symlinks to the `luajit` installation, specifically these (I deleted most of the `ls` output and left just the links):

```
[root@c8db9d10987a ~]# spack cd -i lua-luajit
[root@c8db9d10987a lua-luajit-2.0.5-hr2tpznfx45ryrw2dssblxgscrqi3r7k]# ls -l bin include lib
bin:
lrwxrwxrwx 1 root root     12 Jun  8 00:02 lua -> luajit-2.0.5

include:
lrwxrwxrwx 1 root root   10 Jun  8 00:03 lua -> luajit-2.0

lib:
lrwxrwxrwx 1 root root     22 Jun  8 00:04 liblua.so -> libluajit-5.1.so.2.0.5
```

I think for now we should have the `lua-luajit` installation add those links.  I think we should make a variant for this that causes the links to be created after installation in the `lua-luajit` package.

So:

```python
class LuaLuajit(MakefilePackage):
    ...
    variant('lualinks', default=False, description="add symlinks to make lua-luajit a drop-in lua replacement")

    provides("lua-lang", when="+lualinks")

    @run_after("install")
    @when("+lualinks")
    install_links(self):
        # make symlinks as shown above
```

Then, if the user runs:

```
spack install irep^lua-luajit
```

It will work as you'd expect.  The new concretizer will figure out that it needs to enable `+lualinks` for `lua-luajit` to provide `lua`.  And when it does, it will install the `lua` executable, headers, and libraries that the `irep` build is looking for.

**Review by becker33 — DISMISSED** [review]

**Review by tgamblin — APPROVED** [review]

**Pull request merged**